File: stepik-union/src/main/java/org/stepik/plugin/projectWizard/ui/ProjectSettingsPanel.java
package org.stepik.plugin.projectWizard.ui;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.objects.StudyObject;
import org.stepik.core.SupportedLanguages;
import org.stepik.core.stepik.StepikAuthManager;
import org.stepik.core.stepik.StepikAuthManagerListener;
import org.stepik.core.stepik.StepikAuthState;
import org.stepik.plugin.projectWizard.StepikProjectGenerator;
import org.stepik.plugin.utils.Utils;

import javax.swing.*;
import java.awt.event.HierarchyEvent;
import java.awt.event.HierarchyListener;
import java.util.ArrayList;
import java.util.List;

import static org.stepik.core.stepik.StepikAuthState.AUTH;
import static org.stepik.core.stepik.StepikAuthState.NOT_AUTH;

public class ProjectSettingsPanel implements ProjectSetting, HierarchyListener, StepikAuthManagerListener {
    private static final Logger logger = Logger.getInstance(ProjectSettingsPanel.class);
    private final List<ProjectSettingListener> listeners = new ArrayList<>();
    private JPanel mainPanel;
    private JLabel nameLabel;
    private JLabel userName;
    private JLabel langLabel;
    private LanguageComboBox langComboBox;
    private JLabel courseLabel;
    private CourseListBox courseListComboBox;
    private RefreshButton refreshListButton;
    private CourseDescriptionPane courseListDescription;
    private JScrollPane scrollPane;
    private JButton logoutButton;
    private StudyObject selectedStudyObject = StepikProjectGenerator.EMPTY_STUDY_OBJECT;

    public ProjectSettingsPanel(boolean visibleLangBox) {
        refreshListButton.setTarget(courseListComboBox);
        courseListComboBox.setTarget(this);

        langComboBox.setTarget(this);
        langComboBox.setVisible(visibleLangBox);
        langLabel.setVisible(visibleLangBox);

        mainPanel.addHierarchyListener(this);

        logoutButton.addActionListener(e -> StepikAuthManager.logoutAndAuth());

        StepikAuthManager.addListener(this);
    }

    private void setUsername() {
        String username = StepikAuthManager.getCurrentUserFullName();
        userName.setText(username);
    }

    @NotNull
    public JPanel getComponent() {
        return mainPanel;
    }

    public void updateStep() {
        logger.info("Start updating settings");
        courseListComboBox.refresh(langComboBox.getSelectedItem());
        setUsername();
        logger.info("Updating settings is done");
    }

    public boolean validate() {
        boolean valid = selectedStudyObject.getId() != 0;
        logger.info("Validation is " + valid);
        return valid;
    }

    @Override
    public void selectedStudyNode(@NotNull StudyObject studyObject) {
        selectedStudyObject = studyObject;
        String description = Utils.getCourseDescription(studyObject);
        courseListDescription.setText(description);
        // Scroll to top
        courseListDescription.setSelectionStart(0);
        courseListDescription.setSelectionEnd(0);
        logger.info("Has selected the course: " + studyObject);
        notifyListeners();
    }

    private void notifyListeners() {
        listeners.forEach(ProjectSettingListener::changed);
    }

    @Override
    public void addListener(@NotNull ProjectSettingListener listener) {
        listeners.add(listener);
    }

    @Override
    public void removeListener(@NotNull ProjectSettingListener listener) {
        listeners.remove(listener);
    }

    @Override
    public void selectedProgrammingLanguage(@NotNull SupportedLanguages language) {
        courseListComboBox.refresh(language);
    }

    @NotNull
    public SupportedLanguages getLanguage() {
        return langComboBox.getSelectedItem();
    }

    public void setLanguage(@NotNull SupportedLanguages language) {
        langComboBox.setSelectedItem(language);
    }

    @NotNull
    public StudyObject getSelectedStudyObject() {
        return selectedStudyObject;
    }

    @Override
    public void hierarchyChanged(HierarchyEvent e) {
        notifyListeners();
    }

    @Override
    public void stateChanged(@NotNull StepikAuthState oldState, @NotNull StepikAuthState newState) {
        if (newState == NOT_AUTH || newState == AUTH) {
            ApplicationManager.getApplication()
                    .invokeLater(this::updateStep, ModalityState.stateForComponent(mainPanel));
        }
    }

    public void dispose() {
        StepikAuthManager.removeListener(this);
    }
}

File: stepik-union/src/main/kotlin/org/stepik/plugin/projectWizard/ui/ProjectSettingsPanel.kt
package org.stepik.plugin.projectWizard.ui

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ModalityState
import org.stepik.api.objects.StudyObject
import org.stepik.core.SupportedLanguages
import org.stepik.core.auth.StepikAuthManager
import org.stepik.core.auth.StepikAuthManager.isAuthenticated
import org.stepik.core.auth.StepikAuthManagerListener
import org.stepik.core.auth.StepikAuthState
import org.stepik.core.auth.StepikAuthState.AUTH
import org.stepik.core.auth.StepikAuthState.NOT_AUTH
import org.stepik.core.common.Loggable
import org.stepik.core.utils.EMPTY_STUDY_OBJECT
import org.stepik.core.utils.StepikUrlUtils.getCourseDescription
import java.awt.event.HierarchyEvent
import java.awt.event.HierarchyListener
import java.util.*
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JScrollPane

class ProjectSettingsPanel(visibleLangBox: Boolean) : ProjectSetting, HierarchyListener, StepikAuthManagerListener,
        Loggable {
    
    private val listeners = ArrayList<ProjectSettingListener>()
    private var mainPanel: JPanel? = null
    private var nameLabel: JLabel? = null
    private var userName: JLabel? = null
    private var langLabel: JLabel? = null
    private var langComboBox: LanguageComboBox? = null
    private var courseLabel: JLabel? = null
    private var courseListComboBox: CourseListBox? = null
    private var refreshListButton: RefreshButton? = null
    private var courseListDescription: CourseDescriptionPane? = null
    private var scrollPane: JScrollPane? = null
    private var loginButton: JButton? = null
    var selectedStudyObject = EMPTY_STUDY_OBJECT
        private set
    
    var language: SupportedLanguages
        get() = langComboBox!!.selectedItem
        set(language) {
            langComboBox!!.selectedItem = language
        }
    
    init {
        refreshListButton!!.target = courseListComboBox!!
        courseListComboBox!!.target = this
        langComboBox!!.target = this
        langComboBox!!.isVisible = visibleLangBox
        langLabel!!.isVisible = visibleLangBox
        mainPanel!!.addHierarchyListener(this)
        loginButton!!.addActionListener { StepikAuthManager.relogin() }
        StepikAuthManager.addListener(this)
    }
    
    fun getComponent() = mainPanel
    
    private fun setUsername() {
        val username = StepikAuthManager.currentUserFullName
        userName!!.text = username
    }
    
    fun updateStep() {
        logger.info("Start updating settings")
        courseListComboBox!!.refresh(langComboBox!!.selectedItem)
        setUsername()
        loginButton!!.text = if (isAuthenticated) "Change user" else "Login"
        logger.info("Updating settings is done")
    }
    
    fun validate(): Boolean {
        val valid = selectedStudyObject.id != 0L
        logger.info("Validation is $valid")
        return valid
    }
    
    override fun selectedStudyNode(studyObject: StudyObject) {
        selectedStudyObject = studyObject
        val description = getCourseDescription(studyObject)
        courseListDescription!!.text = description
        // Scroll to top
        courseListDescription!!.selectionStart = 0
        courseListDescription!!.selectionEnd = 0
        logger.info("Has selected the course: $studyObject")
        notifyListeners()
    }
    
    private fun notifyListeners() {
        listeners.forEach { it.changed() }
    }
    
    override fun addListener(listener: ProjectSettingListener) {
        listeners.add(listener)
    }
    
    override fun removeListener(listener: ProjectSettingListener) {
        listeners.remove(listener)
    }
    
    override fun selectedProgrammingLanguage(language: SupportedLanguages) {
        courseListComboBox!!.refresh(language)
    }
    
    override fun hierarchyChanged(e: HierarchyEvent) {
        notifyListeners()
    }
    
    override fun stateChanged(oldState: StepikAuthState, newState: StepikAuthState) {
        if (newState == NOT_AUTH || newState == AUTH) {
            ApplicationManager.getApplication()
                    .invokeLater({ this.updateStep() }, ModalityState.stateForComponent(mainPanel!!))
        }
    }
    
    fun dispose() {
        StepikAuthManager.removeListener(this)
    }
}
